source/common/upstream/upstream_impl.cc

```diff
@@ -890,6 +890,65 @@ createOptions(const envoy::config::cluster::v3::Cluster& config,
       config.has_http2_protocol_options(), validation_visitor);
 }
 
+std::pair<LoadBalancerConfigPtr, TypedLoadBalancerFactory*>
+LegacyLbPolicyConfigHelper::getTypedLbConfigFromLegacyProtoWithoutSubset(
+    const envoy::config::cluster::v3::Cluster& cluster,
+    ProtobufMessage::ValidationVisitor& visitor) {
+
+  LoadBalancerConfigPtr lb_config;
+  TypedLoadBalancerFactory* lb_factory = nullptr;
+
+  switch (cluster.lb_policy()) {
+    PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
+  case envoy::config::cluster::v3::Cluster::ROUND_ROBIN:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.round_robin");
+    break;
+  case envoy::config::cluster::v3::Cluster::LEAST_REQUEST:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.least_request");
+    break;
+  case envoy::config::cluster::v3::Cluster::RANDOM:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.random");
+    break;
+  case envoy::config::cluster::v3::Cluster::RING_HASH:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.ring_hash");
+    break;
+  case envoy::config::cluster::v3::Cluster::MAGLEV:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.maglev");
+    break;
+  case envoy::config::cluster::v3::Cluster::CLUSTER_PROVIDED:
+    lb_factory = Config::Utility::getFactoryByName<TypedLoadBalancerFactory>(
+        "envoy.load_balancing_policies.cluster_provided");
+    break;
+  case envoy::config::cluster::v3::Cluster::LOAD_BALANCING_POLICY_CONFIG:
+    PANIC("getTypedLbConfigFromLegacyProtoWithoutSubset: should not reach here");
+    break;
+  }
+
+  ASSERT(lb_factory != nullptr);
```

Comment:
I don't think we can assert this is non-null - if the factory is not compiled in this will crash immediately below, no?  I think we should gracefully handle failure (either return null/null or make the first a StatusOr

Reply:
sgtm.